src/handle/why3_tactic.ml

```diff
@@ -42,30 +42,84 @@ let why3_env : Why3.Env.env =
-  ; symb_T   : sym (** Encoding of types.        *)
-  ; symb_or  : sym (** Disjunction(∨) symbol.    *)
-  ; symb_and : sym (** Conjunction(∧) symbol.    *)
-  ; symb_imp : sym (** Implication(⇒) symbol.    *)
-  ; symb_bot : sym (** Bot(⊥) symbol.            *)
-  ; symb_top : sym (** Top(⊤) symbol.            *)
-  ; symb_not : sym (** Not(¬) symbol.            *) }
+  { symb_P     : sym (** Encoding of propositions. *)
+  ; symb_T     : sym (** Encoding of types.        *)
+  ; symb_or    : sym (** Disjunction(∨) symbol.    *)
+  ; symb_and   : sym (** Conjunction(∧) symbol.    *)
+  ; symb_imp   : sym (** Implication(⇒) symbol.    *)
+  ; symb_bot   : sym (** Bot(⊥) symbol.            *)
+  ; symb_top   : sym (** Top(⊤) symbol.            *)
+  ; symb_not   : sym (** Not(¬) symbol.            *)
+  ; symb_ex    : sym (** Exists(∃) symbol.         *)
+  ; symb_all   : sym (** Forall(∀) symbol.         *)
+  ; symb_codes : sym (** Type of encoded types.    *) }
 
 (** [get_config ss pos] build the configuration using [ss]. *)
 let get_config : Sig_state.t -> Pos.popt -> config = fun ss pos ->
   let builtin = Builtin.get ss pos in
-  { symb_P   = builtin "P"
-  ; symb_T   = builtin "T"
-  ; symb_or  = builtin "or"
-  ; symb_and = builtin "and"
-  ; symb_imp = builtin "imp"
-  ; symb_bot = builtin "bot"
-  ; symb_top = builtin "top"
-  ; symb_not = builtin "not" }
+  { symb_P     = builtin "P"
+  ; symb_T     = builtin "T"
+  ; symb_or    = builtin "or"
+  ; symb_and   = builtin "and"
+  ; symb_imp   = builtin "imp"
+  ; symb_bot   = builtin "bot"
+  ; symb_top   = builtin "top"
+  ; symb_not   = builtin "not"
+  ; symb_ex    = builtin "ex"
+  ; symb_all   = builtin "all"
+  ; symb_codes = builtin "Set" }
 
 (** A map from lambdapi terms to Why3 constants. *)
 type cnst_table = (term * Why3.Term.lsymbol) list
 
+(** Table of declared types, these types may be variables or constants. *)
+module TyTable : sig 
+  type t 
+
+  val empty : t 
+
+  val ty_of_term : t -> term -> (t * Why3.Ty.ty) option 
+  (** [ty_of_term tbl te] fetches the type associated to term [te] in
+      table [tbl] if it exists, or it creates such a type and return
+      the created type and the new table. *)
+
+  val fold_left : 
+    do_var:('a -> term -> Why3.Ty.tvsymbol -> 'a) ->
+    do_sym:('a -> term -> Why3.Ty.tysymbol -> 'a) -> 'a -> t -> 'a
+  (** [fold_left ~do_var ~do_sym init tbl] folds over the table [tbl]
+      with initial value [init] using [do_var] to operate on type
+      variables and [do_sym] to operate on type symbols. *)
+end = struct
+  (* The module is created to hide the type [sym_or_var] *)
+  type sym_or_var = TySym of Why3.Ty.tysymbol | TyVar of Why3.Ty.tvsymbol
+
+  type t = (term * sym_or_var) list
+
+  let empty = []
+
+  let ty_of_term tbl te =
+    match List.assoc_eq (Eval.eq_modulo []) te tbl with
+    | TySym s -> Some (tbl, Why3.Ty.ty_app s [])
+    | TyVar v -> Some (tbl, Why3.Ty.ty_var v)
+    | exception Not_found -> (
```

Comment:
The parenthesis is not needed here, is it?

Reply:
Indeed they may not be necessary because it's the last clause, but it's a bit error prone because we may not distinguish the inner match from the outer.